### ISSUE TITLE
Wanted: option to completely ignore the heat network and gas network

### ISSUE THREAD

**Issue by ChaelKruip** [quintel/etmoses#1319]
If you don't want a LES without heat infra or gas infra you now need to set all number_of_units and distances to zero. This is annoying and intransparent.

Perhaps we can allow the user to 'disable' the whole infrastructure for heat or gas with a switch like we do with the Merit Order in the ETM?

**Comment by grdw** [comment]
- [ ] Add extra column called 'enabled' to `asset_lists`
- [ ] Add buttons to gas asset list edit page and heat source list edit page
- [ ] Ignore the 'gas load' page and the 'heat load' page when both options are disabled. Set the corresponding business case costs to 0.

**Comment by grdw** [comment]
I am now able to disable the Gas asset list and Heat asset list however I have no idea what 'disable' means in this context. I assumed this was to exclude it from the business case. This was an understatement. 

The following things could be part of 'disable'

- Ignore the business case
- Ignore all gas/heat load (also at end-points)
- Ignore all gas/heat load related market model measures (for production and consumption?)

**Comment by ChaelKruip** [comment]
> [...] however I have no idea what 'disable' means in this context.

Perhaps we can break this discussion up into sections:

* Scope
* What should it look like to the user? / What can the user do?
* What happens to the load calculation?
* What happens to the Market model?
* What happens to the Business case?

## Scope

My feeling is that this functionality is relevant for
* gas infra
* heat infra
* district heat sources

##  What should it look like to the user? / What can the user do?

My feeling is that we should present the user with a clearly visible toggle (similar to Merit in ETM) to 'ignore this infrastructure' / 'include this infrastructure'

@grdw can you provide a screenshot of what you already did for the sake of discussion?

## What happens to the load (calculation)?

This is a difficult one! 

One extreme would be to make the proposed toggle have no effect at all on the load calculations. Although this is simple to implement it could lead to unrealistically cheap LESses: it would allow the user to benefit from deploying gas-technologies / district heating without having to pay for the infrastructure!

I think we should at the very least warn the user when toggling the infra-structure for a carrier to 'off' this is inconsistent with technologies chosen in the technology matrix. 

The other extreme is that we don't allow gas/heat consumption if these infrastructures are toggled 'off'. This is more realistic but possibly a lot of work to implement. @antw @grdw can you give us an estimate of the time required to make the gas-load calculation and heat-load calculation optional?

## What happens to the Market model?

This depends on what we choose for the load calculation. If we don't allow gas/heat consumption, the transactions based on gas/heat consumption will yield zero as well...

## What happens to the Business case?

I think that we should omit all investment and O&M costs of infrastructure for gas/heat from the business cases. 

If we opt to omit the load calculations, the effects of gas/heat-based transactions will make it into the business case as well.


@dennisschoenmakers @antw @grdw @marjoleinschwachofer I'd like to hear your ideas/opinions as well!

**Comment by grdw** [comment]
> scope

>  gas infra
>   heat infra
>   district heat sources

Disabling the heat infra now results in the disabling of district heat sources and the other way around. So I think we should talk about disabling the "heat network" and disabling the "gas network" as a whole.

>  can you give us an estimate of the time required to make the gas-load calculation and heat-load calculation optional?

Right now the `calculation_context` contains all three networks in an array. We might add/remove them depending on which asset list is enabled or not? Could you verify it might work something like that @antw?

**Comment by antw** [comment]
> If you don't want a LES without heat infra or gas infra you now need to set all number_of_units and distances to zero. This is annoying and intransparent.

Can a visitor not just delete the infrastructure from the LES? I think this is opaque only because the business case doesn't explain the breakdown of each cost, which may leave a visitor wondering where a number came from. I think it's *really* unintuitive to see a list of heat and gas assets, only have to them do nothing in the LES because a magic "Disable [carrier]?" button was pressed.

If setting the number of units for these networks is not intuitive, then a magic on/off button hidden on the Edit form is like taking that up to 11. 😢

Alternatively, the import form could contain options similar to the scenario scaling form in the ETM, by allowing the visitor to choose what they'd like to import:

```
Topology:     | Default topology     |▼|
Market model: | Default market model |▼|
Scenario:     | ...                    |

Include:      [✓] Gas infrastructure
              [✓] Heat infrastructure
```

Untick the gas or heat infrastructure and we'll not import anything (i.e. set the relevant asset lists to be `[]`). This should also avoid the need to litter the application with endless `if gas_enabled? ... if heat_enabled? ...`.

A visitor may still choose to add infrastructure later (manually, by hand), and we won't run the risk of bug reports like "my P2G isn't working even with the strategy enabled", or "why isn't any heat being buffered?".

**Comment by grdw** [comment]
> i.e. set the relevant asset lists to be []

Couldn't we just decide to not create the asset lists at all? I mean, it saves database clutter. Or do we allow the user - after he/she created the LES - to manually add technologies to the empty asset lists?

**Comment by ChaelKruip** [comment]
> If setting the number of units for these networks is not intuitive, then a magic on/off button hidden on the Edit form is like taking that up to 11.

Fair point. Like the reference to [http://www.imdb.com/title/tt0088258/](http://www.imdb.com/title/tt0088258/) 😄 

> Alternatively, the import form could contain options similar to the scenario scaling form in the ETM, by allowing the visitor to choose what they'd like to import:

I like this idea a lot! 👍 

Perhaps we should warn the user if he/she un-checks a box but has technologies in the LES that depend on those infrastructures?

> Or do we allow the user - after he/she created the LES - to manually add technologies to the empty asset lists?

I think so yes.

**Comment by antw** [comment]
> Couldn't we just decide to not create the asset lists at all?

We could, but then you have to handle that explicitly with `if les.gas_asset_list.present?` *every* time you want to do something with an asset list. It's very easy to introduce bugs, because one day we'll assume that the assets lists will be there and forget to add the conditional.

Not adding a blank `asset_list` to the DB *would* be okay so long as we have something like:

```
class TestingGround
  def gas_asset_list
    super || build_gas_asset_list
  end
end
```

... that way, the rest of the app can always assume that an asset list is there.

**Comment by grdw** [comment]
> I think so yes. 

This also takes some changes to the current `asset_list` forms, since you can't add a technology to the table if your `asset_list` is blank. If you'd move the "+" icon to the header row this problem would be solved :)

**Comment by ChaelKruip** [comment]
> This also takes some changes to the current asset_list forms, since you can't add a technology to the table if your asset_list is blank. If you'd move the "+" icon to the header row this problem would be solved :)

I am in favour of this anyways because now you can't delete the last line even if you wanted to remove infrastructure manually.

**Comment by ChaelKruip** [comment]
I think we should go ahead with this. Assigning @grdw.

**Comment by grdw** [comment]
![screen shot 2016-08-25 at 12 41 22](https://cloud.githubusercontent.com/assets/2676542/17966332/4125dc12-6ac1-11e6-9bbc-583e07281f05.png)

I already started on changing the asset list table so it's possible to remove every row.

**Comment by grdw** [comment]
There are currently two ways of making the asset table fully removable:

**My first idea:**
![screen shot 2016-08-25 at 14 42 26](https://cloud.githubusercontent.com/assets/2676542/17969417/3c421fba-6ad2-11e6-8257-d4cb8d198e05.png)

*Pro's*
- The table looks empty

*Con's*
- It's pretty unclear what the '+' button will 'add' (you can't control the newly added row).

---

**My second idea**
![screen shot 2016-08-25 at 14 42 56](https://cloud.githubusercontent.com/assets/2676542/17969416/3c406a3a-6ad2-11e6-8854-2f2557975211.png)

*Pro's*
- You can control the newly added row before adding it.

*Con's*
- The table looks as though there's still one row left.

---

I tend to prefer the last idea, simply because it doesn't look so half-assed. The '+' button in the corner looks practical but it doesn't look really nice.

**Comment by ChaelKruip** [comment]
I like the first idea because it conveys the message that the table is empty. 

In the second case, a user might fill in values for the row which is shown and forget to press "+" (why would he/she? The row is already there right?) This would be more confusing than option 1 IMO.

I'd like to hear other opinions as well though! @antw @jorisberkhout @dennisschoenmakers @MarjoleinSchwachofer @ThomBuijs @DorinevanderVlies?

**Comment by antw** [comment]
> Pro's: The table looks empty
> Con's: It's pretty unclear what the '+' button will 'add' (you can't control the newly added row).

If the table is empty, render an "Add heat source" button in place of the empty row:

![addsource](https://cloud.githubusercontent.com/assets/4383/17973528/ba032c8e-6ada-11e6-9713-db01b4d44c56.png)

**Comment by grdw** [comment]
> allowing the visitor to choose what they'd like to import:

If the asset list is empty couldn't we just also remove the header of the table? And ofourse; if you add a new row, show the header.

*Technical questions*
I'm currently trying to set `import_gas_assets` and `import_heat_assets`. However the `check_box` form helper automatically asumes that the value is either '1' or '0'. The `Import` class does not know how to deal with this since the `attr_reader` doesn't know which type the value is (we could fix that by making the `Import` class a `Virtus` object but that seems a little too much stating that it's only for 2 attributes). How would I approach this? Is there some magic helper that let's you send Boolean's over to a non ActiveRecord object?

Also to blank the `GasAssetList`, `HeatAssetList` and `HeatSourceList` I need to temporarily set some values at the `TestingGround`, is that ok to do that in the `TestingGround` or is that not within the scope of the `TestingGround`?

**Comment by grdw** [comment]
Going from one problem to the next. Adding a button to create a new table row is a good idea. However, removing all table rows, actually removes them from the dom. This is 'great' and not so 'great'. Great because gone = gone. Not so great, because the 'Add a must-run heat source' button has no dom-elements to draw from. I'm trying to solve this by abusing the `<tfoot>` element and placing a hidden table-row in there that will serve as the template for the heat sources, gas assets, etc. 

However, this implementation will take some time

Todo:
- [ ] market_model_table
   - [ ] market model table tfoot template
   - [ ] market model table remove button
- [ ] gas_asset_list_table
   - [ ] gas asset table tfoot template
   - [ ] gas asset table remove button
- [ ] heat_source_list_table
   - [ ] heat source list must run table tfoot template
   - [ ] heat source list must run remove button
   - [ ] heat source list dispatchable table tfoot template
   - [ ] heat source list dispatchable remove button
- [ ]  heat_asset_list_table
   - [ ] heat source list secondary table tfoot template
   - [ ] heat source list secondary remove button

**Comment by antw** [comment]
> I need to temporarily set some values at the TestingGround, is that ok to do that in the TestingGround or is that not within the scope of the TestingGround?

Its a little ugly to store temporary data with the TestingGround, since it means adding columns which will never be used except during the import.

I see two ways around this:

* Add a new `imports` table and model which stores information about the import:

  * `id`
  * `scenario_id`
  * `topology_id`
  * `market_model_id`
  * **`gas_enabled`** (bool)
  * **`heat_enabled`** (bool)

  ... then when the visitor submits [this form](https://beta-moses.energytransitionmodel.com/testing_grounds/import) we create a new Import record and redirect them to `/imports/:id` where they can customise their technology profile (from that point onwards the process is much the same as it is today).

  As a bonus, perhaps the `0`/`1` problem is also fixed by creating a model?

* Alternatively, the temporary values could be stored in [the session](http://guides.rubyonrails.org/action_controller_overview.html#session). This would be much easier.

  Personally I'm not a fan of using the session since it feels like it makes the import rely on external state (i.e. the way the import behaves [heat or no heat, gas or no gas] depends on "magic" variables in the controller).

I think there's a lot of merit to creating a full `Import` record and `ImportsController`; it's a bit more work than the alternative, but it has the benefit of also cleaning up TestingGroundsController by relocating all the actions related to imports.

**Comment by grdw** [comment]
> since it means adding columns which will never be used except during the import.

This is the same when adding a separate `import` table and putting those columns in there. So I don't see any validity in the `imports` table.

Maybe we could set the two checkboxes through [form objects](https://robots.thoughtbot.com/activemodel-form-objects). This problem sounds exactly like the right use for it. So instead of using the `@testing_ground` as the actual form. Use a form object.

**Comment by antw** [comment]
> This is the same when adding a separate import table and putting those columns in there. So I don't see any validity in the imports table.

Not really. There's a difference between a *table* which stores temporary data – which can be cleaned out regularly if necessary, without side-effects – and adding one-time-use columns to an unrelated table that forever take up space (remember also that ActiveRecord reads and writes all those values every time you `find` or `save`). A permanent record of the import has the advantage of repeatability: you can visit `/imports/:id` many times to re-import the same scenario/settings.

Nonetheless, a form object sounds like a simpler alternative.

**Comment by grdw** [comment]
> Nonetheless, a form object sounds like a simpler alternative.

I tried this yesterday for a moment. But this sounded nice on paper, but hard todo in practise. The `Import` form needs to send all the data to the `#new` (profile matrix) form and not build an actual `TestingGround` (which is a bit of a large step). 

Than the `form_for` at the new page had to change to the form object which was harder than it looked considering the major use of helpers and checking things based on the pre-build `TestingGround` in the import class.

I have no idea how much time it would cost if I had continued. 

### Form object option

- [ ] Create a form object for `#new`
- [ ] Link the import to the form object
- [ ] Update the `create` method in the TestingGrounds controller with the new form object (maybe get rid of the `TestingGround::Creator` while this is being done.
- [ ] Don't destroy `general_form` and `technologies_form` in the process since they're also used at the `edit` page


### Import table option

- [ ] Create a new table (With 5 columns)
- [ ] Remove the other columns from the LES (make an association between Import and TestingGround) ?
- [ ] Migrate old LES's?
- [ ] Update the form for the TestingGrounds#import to -> Imports#new
- [ ] Submit the form to Imports#create instead of TestingGrounds#perform_import
- [ ] Imports#create should render TestingGrounds#new

---

I'm a bit uncertain when it comes to making time estimates for both these options but considering the wasteland we could be creating from "Form object option", I'd prefer the Import table option. However, if that takes twice as long... So it depends a little on how much work is it really.
